online-shop/src/app/components/containers/product-form/product-form.component.ts

```diff
@@ -0,0 +1,113 @@
+import {
+  Component,
+  Input,
+  OnChanges,
+  OnInit,
+  SimpleChanges,
+} from '@angular/core';
+import { ActivatedRoute } from '@angular/router';
+import { ProductService } from 'src/app/services/products.service';
+import { Product } from 'src/app/modules/shared/types/products.types';
+import {
+  FormGroup,
+  FormBuilder,
+  Validators,
+  FormControl,
+} from '@angular/forms';
+@Component({
+  selector: 'app-product-form',
+  templateUrl: './product-form.component.html',
+  styleUrls: ['./product-form.component.scss'],
+})
+export class ProductFormComponent implements OnInit {
+  @Input() productId: string | null | undefined;
+  @Input() mode!: string;
+  productForm!: FormGroup;
+  product!: Product;
+  concatenatedString!: string;
+
+  constructor(
+    private route: ActivatedRoute,
+    private productService: ProductService,
+    private formBuilder: FormBuilder
+  ) {}
+
+  ngOnInit() {
+    this.productForm = new FormGroup({
+      name: new FormControl('', Validators.required),
+      category: new FormControl('', Validators.required),
+      price: new FormControl(0, [Validators.required, Validators.min(0)]),
+      image: new FormControl('', Validators.required),
+      description: new FormControl('', Validators.required),
+    });
+    console.log(this.productId);
+    if (this.productId) {
+      this.productService.getProduct(this.productId).subscribe((product) => {
```

Comment:
Make sure you unsubscribe from the subscription

Reply:
Used @UntilDestroy({ checkProperties: true }) for every subscription.